### ISSUE TITLE
SparseDirectUMFPACK does not work correctly with BlockSparseMatrix

### ISSUE THREAD

**Issue by kuljitvirk** [dealii/dealii#12916]
Observe that at line 202 in sparse_direct.cc, the indexing variable should be `element` not `cursor`. Because of this line ALL BlockSparseMatrix based codes in complex domain will give wrong results when using UMFPACK. Please fix this in upcoming releases. Some of us use this library for doing electrodynamics with complex media and it was a long frustrating journey to this bug.

Here is your code around this line
```
   for (size_type row = 0; row < matrix.m(); ++row)
     {
       long int cursor = Ap[row];
       for (size_type block = 0; block < matrix.n_block_cols(); ++block)
         {
           // find the next out-of-order element
           while ((cursor < Ap[row + 1] - 1) && (Ai[cursor] < Ai[cursor + 1]))
             ++cursor;
 
           // if there is none, then just go on
           if (cursor == Ap[row + 1] - 1)
             break;
 
           // otherwise swap this entry with successive ones as long as
           // necessary
           long int element = cursor;
           while ((element < Ap[row + 1] - 1) && (Ai[element] > Ai[element + 1]))
             {
               std::swap(Ai[element], Ai[element + 1]);
 
               std::swap(Ax[element], Ax[element + 1]);
               if (numbers::NumberTraits<number>::is_complex == true)
                 std::swap(Az[cursor], Az[cursor + 1]); // THE BUG
 
               ++element;
             }
         }
     }
 }
```

**Comment by bangerth** [comment]
Ah yes, this looks wrong :-( Good detective work finding this place! 

Could we interest you in creating a patch? You'd then also be listed among the authors of deal.II!

**Comment by kuljitvirk** [comment]
Hi Wolfgang,

Thanks for quick response. I would love to create a patch and contribute
examples after validating them. I actually have never made a patch myself
and so would appreciate instructions.

Thanks
Kuljit

**Comment by jppelteret** [comment]
Hi @kuljitvirk,

You can find some basic information [here, on our Wiki page](https://github.com/dealii/dealii/wiki/Contributing). At the bottom of the page is some additional information on how to use Git and GitHub.

**Comment by bangerth** [comment]
@kuljitvirk Do you need any help with writing a patch for this?

**Comment by kuljitvirk** [comment]
Hi Wolfgang,

Sorry, I had been away from this project. It appears that I just have to
check out deal.ii into my github and then push changes up to it. After that
I will open a pull request. I will do it by 12/8.

Thanks
Kuljit

**Comment by bangerth** [comment]
Excellent, we look forward to it!

**Comment by bangerth** [comment]
Fixed by #13095.
